File: 1/2021.08.24/787.Cheapest Flights Within K Stops/java/solution.java
class Solution {
    public int findCheapestPrice(int n, int[][] flights, int src, int dst, int k) {
        int[] dp = new int[n];
        Arrays.fill(dp, -1);
        dp[src] = 0;
        int res = -1;
        for (int t = 1; t < k + 2; t++) {
            int[] ndp = new int[n];
            Arrays.fill(ndp, -1);
            for (int[] flight : flights) {
                int j = flight[0], i = flight[1], cost = flight[2];
                if (dp[j] != -1) {
                    if (ndp[i] == -1) {
                        ndp[i] = dp[j] + cost;
                    } else {
                        ndp[i] = Math.min(ndp[i], dp[j] + cost);
                    }
                }
            }
            dp = ndp;
            if (res == -1) {
                res = dp[dst];
            } else if (dp[dst] != -1){
                res = Math.min(dp[dst], res);
            }
        }
        return res;
    }
}

File: 1/2021.08.24/787.Cheapest Flights Within K Stops/python/solution.py
class Solution:
    def findCheapestPrice(self, n: int, flights: List[List[int]], src: int, dst: int, k: int) -> int:
        dp = [float('inf')] * n
        res = float('inf')
        dp[src] = 0
        for _ in range(1, k + 2):
            ndp = [float('inf')] * n
            for j, i, cost in flights:
                ndp[i] = min(ndp[i], dp[j] + cost)
            res = min(res, ndp[dst])
            dp = ndp
        return -1 if res == float('inf') else res
